File: 43_prefix_sums/43_1_channel_views.py
def channel_views(views: list, periods: list)->list:
  prefix_sums = []
  output = []
  previous_val = 0
  for view in views:
    previous_val += view
    prefix_sums.append(previous_val)
  
  for period in periods:
    start = period[0]
    end = period[1]
    if start == 0:
      output.append(prefix_sums[end])
    else:
      output.append(prefix_sums[end] - prefix_sums[start-1])
  
  return output

if __name__ == "__main__":
  views = [3, 5, 4, 8, 7, 2, 5, 3, 2, 3]
  periods = [[0, 1], [0, 5], [5, 8], [3, 3]]
  print(channel_views(views, periods))

File: 43_prefix_sums/43_1_channel_views.cpp
#include <iostream>
#include <string>
#include <vector>

std::vector<int> channel_views(std::vector<int>& views, std::vector<std::vector<int>>& periods) {
  std::vector<int> prefix_sums;
  std::vector<int> output;
  int previous_val = 0;

  for (int view: views) {
    previous_val += view;
    prefix_sums.push_back(previous_val);
  }

  for (const std::vector<int>& period: periods) {
    int start = period[0];
    int end = period[1];
    if (start == 0) {
      output.push_back(prefix_sums[end]);
    }else {
      output.push_back(prefix_sums[end] - prefix_sums[start-1]);
    }
  }

  return output;
}

int main() {
  std::vector<int> views = {3, 5, 4, 8, 7, 2, 5, 3, 2, 3};
  std::vector<std::vector<int>> periods = {{0, 1}, {0, 5}, {5, 8}, {3, 3}};

  std::vector<int> result = channel_views(views, periods);

  std::cout << "[";
  for (size_t i = 0; i < result.size(); i++) {
    std::cout << result[i];
    if (i < result.size() - 1) {
      std::cout << ", ";
    }
  }
  std::cout << "]" << std::endl;

  return 0;
}

